### DISCUSSION TITLE
recommended partition num when using partition key based multi tenancy

### DISCUSSION THREAD

**Discussion by rere950303** [milvus-io/milvus#33811] [Q&A and General discussion] [answered]
Among the multi-tenancy strategies, I think the tenancy will be over 10,000, so I would like to use the partition key based method. At this time, I'm thinking about how many partition nums I should use. There are 64 defaults and i can make up to 4096, so do you have any recommendations? I hope the following are considered.

1. I hope there is less conflict between the hash value of the field corresponding to the partition key. I think that will increase the performance.
2. I hope the data channel is considered. I understand that too many partition nums can degrade the overall performance of the system by the data channel time tick mechanism.

**Comment by yhmo** [comment, marked as answer]
If you only have one collection in the server, you can use more partitions.
Assume there are 10 million vectors from 10000 tenancies, 1000 vectors per tenancy.
(1) If the partition number is 100 ----------------- 100 tenancies per partition, 100k vectors per partition
(2) If the partition number is 1000 ---------------  10 tenancies per partition, 10k vectors per partition

Let's say the index is "FALT", brute-force search.
When you search with the expression "tenancy == XX", milvus hashes the "XX" to a partition, and searches data in this partition.
For (1), milvus filer the 100k vectors with the expression to construct a bitset, finds out the 1000 vectors from the 100k vectors, and computes similarity for the 1000 vectors.
For (2), milvus filer the 100k vectors with the expression to construct a bitset, finds out the 1000 vectors from the 10k vectors, and computes similarity for the 1000 vectors.

What is the difference? The difference is "finds out the 1000 vectors from the 100k vectors" and "finds out the 1000 vectors from the 10k vectors". Usually, the filtering stage is not the bottleneck of a search. The performance of (2) is a bit better than (1), a tiny gap.
So, in my opinion, a small number of partitions is better, since a large number cannot get a big improvement in performance. Default value 64 is also ok. Increasing it to 100 is also ok.

**Reply by rere950303** [reply]
@yhmo thank you for reply.
The range filtered by the boolean expr is limited to entities that exist in the partition that corresponds to the hash, right?

> For (2), milvus filer the 100k vectors with the expression to construct a bitset, finds out the 1000 vectors from the 10k vectors, and computes similarity for the 1000 vectors.

here you intended 'For (2), milvus filer the 10k vectors with the expression to construct a bitset' right??

**Reply by yhmo** [reply]
Partition key only benifits the case such as "tenancy == XX" and "tenancy in [XX, YY, ZZ]", but it doesn't help filtering like "tenancy > XX or tenancy < YY".

It is a tradeoff. Too many partitions is not good, brings high pressure to the entire system. Few partitions, too many tanancies in one partition, increase filtering time cost. There might be a "best number", but I don't know, so I will say 100 ~ 200 is ok for the partitions number.
